File: JavaScript/File.js
#! /usr/bin/env node

/*
	このファイルに書かれている内容は Node.js のファイルシステムの内容であり, Filesystem API とは異なる。
	Web上ではテストできない内容である。
*/

const fs = require("fs");
const os = require("os");
const process = require("process");
const child_process = require("child_process");

console.log("\r\nこれからファイル操作を試します\r\n");

console.log("> カレントディレクトリをホームに変更\r\n");
process.chdir(os.homedir());

/*
	Node.js ではファイル操作をする関数が同期処理用と非同期処理用の2つが用意されている。
	非同期処理用のものは,その関数を実行した直後の行から,関数内の処理内容と並列的に進行する。関数を実行し終えた後に実行したい内容は,関数にコールバック関数として渡す。
	同期処理用のものは,その関数を実行した時点で,その直後の行以降の内容は一旦中断される。関数を実行し終えてから直後の行から実行を再開する。
	e.g.
		mkdir(path,callbackFn) : 非同期処理用関数
		mkdirSync(path) : 同期処理用関数
*/

console.log("> Emptyという空フォルダを作成");
fs.mkdirSync("Empty");

console.log("> Blankという空ファイルを作成");
fs.writeFileSync("Blank","");
	// writeFileは既にファイルが存在していれば置換する。

console.log("> フォルダEmptyの中にファイルBlankのシンボリックリンクSymlinkを作成");
fs.symlinkSync("Blank","Empty/Symlink");
	// fs.linkSync("Blank","Empty/Symlink") とすると,ハードリンクが作成される
	/*
		fs.symlinkSync(target,dst,type)
		typeはWindowsのみ影響のある引数 (他のOSでは無視される)
		"file","dir","junction" のうちから選択する
		"file","dir"の場合は引数がなくてもtarget引数の指し示すものから自動的に判別する
	*/

console.log("\r\n");

// 書込み

console.log("> Untitled.mdというMarkdownファイルを作成して書込み");
fs.writeFileSync("Untitled.md","# Header 1");

console.log("\r\n");

// 移動/名称変更

console.log("> フォルダEmptyをPackageに名称変更");
fs.renameSync("Empty","Package");
console.log("> Packageフォルダ内のSymlinkファイルをAliasに名称変更");
fs.renameSync("Package/Symlink","Package/Alias");
console.log("> Untitled.mdを移動して,名称変更");
fs.renameSync("Untitled.md","Package/Headers.md");

console.log("\r\n");

// 追記

console.log("> Markdownファイルに追記");
fs.appendFileSync("Package/Headers.md","\r\n## Header 2\r\n### Header 3");

console.log("\r\n");

// 読込み

console.log("> Markdownファイルを読込み");
console.log("\r\n");
console.log(fs.readFileSync("Package/Headers.md","utf-8"));
	// オプションで "utf-8" を指定しないとバッファデータが返される

console.log("\r\n");

// 再帰的にフォルダ作成

console.log("> フォルダを一気に作成");
fs.mkdirSync("Package/Module/Submodule/Item",{recursive:true});
	// mkdirSyncのオプションで recursive:true とすることで,作成するフォルダItemの上位フォルダModule,Submoduleが存在していなくても,同時に生成される
	// 他のオプションとして,フォルダの権限を指定するmodeもある

console.log("\r\n");

// ファイル/フォルダの複製

console.log("> BlankとModuleを複製");
fs.copyFileSync("Blank","Package/Blank");
	// フォルダの複製は Node.js 標準のモジュールだけではできない

console.log("\r\n");

// ファイル/フォルダの削除

console.log("> ファイルBlankを削除");
fs.unlinkSync("Blank");

console.log("> シンボリックリンクAliasを削除");
fs.unlinkSync("Package/Alias");

// console.log("> フォルダModuleを削除");
// fs.rmdirSync("Package/Module");
	// フォルダをunlinkSyncで削除することはできない
	// 逆に,ファイルやシンボリックリンクをrmdirSyncで削除することもできない
	// フォルダは空の場合のみfs.rmdirSyncで削除できる
	// 空でないフォルダは Node.js 標準モジュールでは削除できない

console.log("\r\n");

check=pt=>{
	if (fs.existsSync(pt)) console.log("   存在しています");
	else console.log("   存在していません");

	if (fs.existsSync(pt)) {
		stat=fs.statSync(pt);
		if (stat.isSymbolicLink()) {
			console.log("   シンボリックリンクです");
			console.log(`   リンク先: ${fs.readlinkSync(pt)}`);
			if (stat.isFile()) console.log("   リンク先はファイルです");
			if (stat.isDirectory()) console.log("   リンク先はフォルダです");
		}
		else {
			if (stat.isFile()) console.log("   ファイルです");
			if (stat.isDirectory()) console.log("   フォルダです");
		}
		if (stat.mode & fs.constants.S_IRUSR) console.log("   読込可能です");
		if (stat.mode & fs.constants.S_IWUSR) console.log("   書込可能です");
		if (stat.mode & fs.constants.S_IXUSR) console.log("   実行可能です");
	}
};

console.log("> 実行ファイルを作成します");
fs.writeFileSync("Package/JavaScript.js",`#! /usr/bin/env node

console.log("Hello world!");
`);

console.log("> このファイルに対する現在の状態を確認");
check("Package/JavaScript.js");

console.log("> 読込/書込権限を剥奪");
fs.chmodSync("Package/JavaScript.js",0o044);

console.log("> 状態を確認");
check("Package/JavaScript.js");

console.log("> 読込/実行権限を付加");
fs.chmodSync("Package/JavaScript.js",0o544);

console.log("> 状態を確認");
check("Package/JavaScript.js");

/*
	fs.chmodSyncは数値により権限を指定する
		4: 読込可能 (Readable)
		2: 書込可能 (Writable)
		1: 実行可能 (eXecutable)
	これらの和を3つ並べて指定する。1つ目はユーザの権限,2つ目はゲストの権限,3つ目はその他の者の権限である。
	JavaScriptでは8進数で表記するために,前に0oを付加する
	e.g. fs.chmodSync(somefile,0o754)
		ユーザは 7 = 4+2+1 だから,読込可能,書込可能,実行可能
		ゲストは 5 = 4 + 1 だから,読込可能,実行可能
		その他は 4 だから,読込のみ可能
*/

console.log("> 実行");
rtn=child_process.execSync("Package/JavaScript.js").toString();
	// execSyncの戻り値は標準出力のデータのバッファである。toStringで文字列化する。
console.log(rtn);

console.log("\r\n");

File: Julia/Legacy/File.jl
#! /usr/bin/env julia

println("\r\nこれからファイル操作を試します\r\n")

println("> カレントディレクトリをホームに変更\r\n")
cd(homedir())

# ファイル/フォルダ/シンボリックリンクの作成

println("> Emptyという空フォルダを作成")
mkdir("Empty")

println("> Blankという空ファイルを作成")
touch("Blank")

println("> フォルダEmptyの中にファイルBlankのシンボリックリンクSymlinkを作成\r\n")
symlink("Blank","Empty/Symlink")

# 書込み

print("> Untitled.mdというMarkdownファイルを作成して書込み\r\n")
io=open("Untitled.md","w")
write(io,"# Header 1")
close(io)

# 移動/名称変更

println("> フォルダEmptyをPackageに名称変更")
mv("Empty","Package")
println("> Packageフォルダ内のSymlinkファイルをAliasに名称変更")
mv("Package/Symlink","Package/Alias")
println("> Untitled.mdを移動して,名称変更\r\n")
mv("Untitled.md","Package/Headers.md")

# 追記

println("> Markdownファイルに追記\r\n")
io=open("Package/Headers.md","a")
write(io,"\r\n## Header 2\r\n### Header 3")
close(io)

# 読込み

println("> Markdownファイルを読込み\r\n")
io=open("Package/Headers.md","r")
print(read(io,String))
close(io)
println("\r\n")

# 再帰的にフォルダ作成

println("> フォルダを一気に作成\r\n")
mkpath("Package/Module/Submodule/Item")

# ファイル/フォルダの複製

println("> BlankとModuleを複製\r\n")
cp("Blank","Package/Blank")
cp("Package/Module","Package/Module Copy")

# ファイル/フォルダの削除

println("> ファイルBlankを削除")
rm("Blank")

println("> シンボリックリンクAliasを削除")
rm("Package/Alias")

println("> フォルダModuleを削除\r\n")
rm("Package/Module",recursive=true)
	# 空のフォルダであればそのまま削除できる。空でない場合は削除できない。
	# recursive=true を追加すれば,空でなくても削除できる
	# force=true を追加すれば,ファイル/フォルダが実在しない場所を指定してもエラーを返さない

# 状態を確認する関数

function check(pt::String)
	if islink(pt)
		println("   シンボリックリンクです")
		println("   リンク先: $(readlink(pt))")
		if isfile(pt)
			println("   リンク先はファイルです")
		end
		if isdir(pt)
			print("   リンク先はフォルダです")
		end
	else
		if isfile(pt)
			println("   ファイルです")
		end
		if isdir(pt)
			println("   フォルダです")
		end
	end
	if uperm(pt)&04>0
		println("   読込可能です")
	end
	if uperm(pt)&02>0
		println("   書込可能です")
	end
	if uperm(pt)&01>0
		println("   実行可能です")
	end
end

# 権限を確認/変更

println("> 実行ファイルを作成します")
io=open("Package/Julia.jl","w")
write(io,"""
#! /usr/bin/env julia

println("Hello world!")
""")
close(io)

println("> このファイルに対する現在の状態を確認")
check("Package/Julia.jl")

println("> 読込/書込権限を剥奪")
chmod("Package/Julia.jl",0o044)

println("> 状態を確認")
check("Package/Julia.jl")

println("> 読込/実行権限を付加")
chmod("Package/Julia.jl",0o544)

println("> 状態を確認")
check("Package/Julia.jl")

#=
	chmodは数値により権限を指定する
		4: 読込可能 (Readable)
		2: 書込可能 (Writable)
		1: 実行可能 (eXecutable)
	これらの和を3つ並べて指定する。1つ目はユーザの権限,2つ目はゲストの権限,3つ目はその他の者の権限である。
	Juliaでは8進数で表記するために,前に0oを付加する
	e.g. chmod(somefile,0754)
		ユーザは 7 = 4+2+1 だから,読込可能,書込可能,実行可能
		ゲストは 5 = 4 + 1 だから,読込可能,実行可能
		その他は 4 だから,読込のみ可能
=#

println("> 実行")
run(`Package/Julia.jl`)
	# `Package/Julia.jl` によりプロセスオブジェクトを生成し,runによりプロセスを実行する

println("\r\n")